agent/agent.go

```diff
@@ -205,6 +209,154 @@ func (a *agent) runLoop(ctx context.Context) {
 	}
 }
 
+func (a *agent) collectMetadata(ctx context.Context, md codersdk.WorkspaceAgentMetadataDescription) *codersdk.WorkspaceAgentMetadataResult {
+	timeout := md.Timeout
+	if timeout == 0 {
+		timeout = md.Interval
+	}
+	ctx, cancel := context.WithTimeout(ctx,
+		time.Duration(timeout)*time.Second,
+	)
+	defer cancel()
+
+	var out bytes.Buffer
+
+	result := &codersdk.WorkspaceAgentMetadataResult{
+		// CollectedAt is set here for testing purposes and overrode by
+		// the server to the time the server received the result to protect
+		// against clock skew.
+		//
+		// In the future, the server may accept the timestamp from the agent
+		// if it is certain the clocks are in sync.
+		CollectedAt: time.Now(),
+	}
+	cmd, err := a.createCommand(ctx, md.Script, nil)
+	if err != nil {
+		result.Error = err.Error()
+		return result
+	}
+
+	cmd.Stdout = &out
+	cmd.Stderr = &out
+
+	// The error isn't mutually exclusive with useful output.
+	err = cmd.Run()
+
+	const bufLimit = 10 << 10
+	if out.Len() > bufLimit {
+		err = errors.Join(
+			err,
+			xerrors.Errorf("output truncated from %v to %v bytes", out.Len(), bufLimit),
+		)
+		out.Truncate(bufLimit)
+	}
+
+	if err != nil {
+		result.Error = err.Error()
+	}
+	result.Value = out.String()
+	return result
+}
+
+func convertInterval(i int64) time.Duration {
```

Comment:
This function is named poorly. I'd recommend `adjustIntervalForTests` or something more verbose. Converting an interval from an integer to a time.Duration wouldn't normally require its own function.